### PR TITLE
Smart update of metric collection

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/metrics/issues/143

Update `MetricCollection` such that metrics that have similar metric states (`Precision` and `Recall` for example) we only calculate the metric state for one of them and then broadcast it to all other metrics with similar metric states.
After the first call to update we checks if the metric states matches between any metrics. If they do we merge them into a compute group of metrics where we will only update the first and broadcast the state to the remaining metrics in the group

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/__init__.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 import logging as __logging
 import os
 
-from torchmetrics.__about__ import *  # noqa: F401, F403
+from torchmetrics.__about__ import *  # noqa: F403
```
Comment (SkafteNicki) [inline torchmetrics/__init__.py:5]:
```suggestion
from torchmetrics.__about__ import * # noqa: F401, F403
```

File: torchmetrics/__init__.py [hunk 2]
```diff
@@ -25,7 +25,7 @@
     SignalDistortionRatio,
     SignalNoiseRatio,
 )
-from torchmetrics.classification import (  # noqa: E402, F401
+from torchmetrics.classification import (  # noqa: E402
```
Comment (SkafteNicki) [inline torchmetrics/__init__.py:28]:
```suggestion
from torchmetrics.classification import ( # noqa: E402, F401
```

File: torchmetrics/__init__.py [hunk 3]
```diff
@@ -185,6 +196,7 @@
     "StatScores",
     "SumMetric",
     "SymmetricMeanAbsolutePercentageError",
+    "WER",
```
Comment (SkafteNicki) [inline torchmetrics/__init__.py:199]:
```suggestion
```

File: torchmetrics/collections.py [hunk 4]
```diff
@@ -117,11 +125,83 @@ def update(self, *args: Any, **kwargs: Any) -> None:
...
+            state2 = getattr(metric2, key)
+
+            if type(state1) != type(state2):
+                return False
+
+            if isinstance(state1, Tensor) and isinstance(state2, Tensor):
+                if state1.shape != state2.shape:
+                    return False
+                if not torch.allclose(state1, state2):
+                    return False
```
Comment (justusschock) [inline torchmetrics/collections.py:193]:
```suggestion
                return state1.shape == state2.shape and torch.allclose(state1, state2):
```

File: torchmetrics/collections.py [hunk 5]
```diff
@@ -117,11 +125,83 @@ def update(self, *args: Any, **kwargs: Any) -> None:
...
+
+            if isinstance(state1, list) and isinstance(state2, list):
+                if any(s1.shape != s2.shape for s1, s2 in zip(state1, state2)):
+                    return False
+                if any(torch.allclose(s1, s2) for s1, s2 in zip(state1, state2)):
```
Comment (justusschock) [inline torchmetrics/collections.py:198]:
```suggestion
                if not all(torch.allclose(s1, s2) for s1, s2 in zip(state1, state2)):
```

When moving it to one line as suggested below, this will be outdated anyways, but currently this behaviour is wrong I think :) It either should be `any(not torch.allclose(...))` or `not all(torch.allclose(...))`

File: torchmetrics/collections.py [hunk 6]
```diff
@@ -117,11 +125,83 @@ def update(self, *args: Any, **kwargs: Any) -> None:
...
+                    return False
```
Comment (justusschock) [inline torchmetrics/collections.py:199]:
```suggestion
                return all(s1.shape == s2.shape and torch.allclose(s1, s2) for s1, s2 in zip(state1, state2))
```

**Comment by SkafteNicki** [comment]
Based on the following script:
```python
from torchmetrics import Accuracy, Precision, Recall, MetricCollection
import torch
import time
import numpy as np

def experiment(enable, steps):

    m = MetricCollection([Accuracy(num_classes=10), Precision(num_classes=10), Recall(num_classes=10)], enable_compute_groups=enable)

    for _ in range(steps):
        preds = torch.randn(256, 10).softmax(dim=-1)
        target = torch.randint(10, (256,))
        m.update(preds, target)

    m.compute()

times = [ ]
for _ in range(5):
    start = time.time()
    experiment(True)
    times.append(time.time() - start)

times2 = [ ]
for _ in range(5):
    start = time.time()
    experiment(False)
    times2.append(time.time() - start)

print(f"{np.mean(times)}+-{np.std(times)}")
print(f"{np.mean(times2)}+-{np.std(times)}")
```
I get 
```
0.42464628219604494+-0.6570260998311238
0.2263005256652832+-0.6570260998311238
```
for `steps=100` and
```
8.417768621444703+-1.1484917218472177
18.352816343307495+-1.1484917218472177
```
for `steps=10000`. The dynamical creating of the compute groups are an expensive search, and therefore not worth it if the user is only doing a few steps. However, as the number of steps increase it will be mush faster. In the example we have three metrics that share a metric state and potentially the new implementation can therefore be close to 3x times faster when the number of steps are large enough.

**Comment by justusschock** [comment]
@SkafteNicki: What about also adding an option for the user to provide compute_groups manually? Something like `compute_groups = False | "auto" | [[0, 1, 2], [3,4]] | [["accuracy", "f1"], ["precision", "recall]]`?

File: torchmetrics/collections.py [hunk 7]
```diff
@@ -45,6 +44,13 @@ class name as key for the output dict.
 
         postfix: a string to append after the keys of the output dict
 
+        enable_compute_groups:
+            By defualt the MetricCollection will try to reduce the computations needed for the metrics in the collection
```
Comment (justusschock) [inline torchmetrics/collections.py:48]:
```suggestion
            By default the MetricCollection will try to reduce the computations needed for the metrics in the collection
```

**Comment by SkafteNicki** [comment]
> @SkafteNicki: What about also adding an option for the user to provide compute_groups manually? Something like `compute_groups = False | "auto" | [[0, 1, 2], [3,4]] | [["accuracy", "f1"], ["precision", "recall]]` for disabling, dynamic search, when using with a Sequence and with a dict?
> 
> That way we would only do some sanity checks (should be a lot cheaper than exhaustive search) and avoid the search at all.
> 

Sounds like a good idea. I implement that :]

> Also you check every metric against every other. But you only need to check against one metric in each group, right?

Actually not every metric is run against every other. I already only compare the first in each group to the first in any other group:
https://github.com/PyTorchLightning/metrics/blob/3ada9afb7d4948ceabc9c06aa862dfed58b24e12/torchmetrics/collections.py#L160-L161

File: torchmetrics/collections.py [hunk 8]
```diff
@@ -45,6 +44,14 @@ class name as key for the output dict.
 
         postfix: a string to append after the keys of the output dict
 
+        compute_groups:
+            By defualt the MetricCollection will try to reduce the computations needed for the metrics in the collection
```
Comment (justusschock) [inline torchmetrics/collections.py:48]:
```suggestion
            By default the MetricCollection will try to reduce the computations needed for the metrics in the collection
```

again :D

**Review by justusschock — COMMENTED** [review]
Just a nit :)

@SkafteNicki did this yield the expected performance boost?

File: torchmetrics/collections.py (hunk 8 above)
Comment (Borda) [reply torchmetrics/collections.py:48]: I see how you enjoy it in your office - just find anything :D

**Comment by Borda** [comment]
> @SkafteNicki did this yield the expected performance boost?

would be cool if we could mention any benchmark on docs to give user reason to use it :rabbit:

File: torchmetrics/collections.py (hunk 8 above)
Comment (justusschock) [reply torchmetrics/collections.py]: I just looked through it again since it popped up in my notifications, but yes, I quite enjoyed it when I saw this :P

File: docs/source/pages/overview.rst [hunk 9]
```diff
@@ -292,6 +292,14 @@ inside your LightningModule
     have the same call signature. If this is not the case, input that should be
     given to different metrics can given as keyword arguments to the collection.
 
+An additional advantage of using the ``MetricCollection`` object is that it will
+automatically try to reduce the computations needed by finding groups of metrics
+that share the same underlying metric state. If such metrics are found only one
+of them are actually updated and the updated state will be broadcastet to the rest
```
Comment (justusschock) [inline docs/source/pages/overview.rst:298]:
```suggestion
of them is actually updated and the updated state will be broadcasted to the rest
```

File: docs/source/pages/overview.rst [hunk 10]
```diff
@@ -292,6 +292,14 @@ inside your LightningModule
...
+of the metrics. In the example above, this will lead to a 2x-3x lower computational
+cost compared to disabling this feature. See the `compute_groups` argument
+for more info.
```
Comment (justusschock) [inline docs/source/pages/overview.rst:301]:
```suggestion
cost compared to disabling this feature. 
However, this speedup comes with a high cost upfront, where the state-groups have to be determined after the first update. This cost can be significant for very low steps (approx. up to 100) but still leads to an overall speedup for everything beyond that. In case the groups are known beforehand, these can also be set manually to avoid the cost of the dynamic search.
See the :param:`torchmetrics.MetricCollection.compute_groups` argument for more information on this topic.
```

**Review by justusschock — APPROVED** [review]
LGTM :) added some more text on potential slowdowns

File: CHANGELOG.md [hunk 11]
```diff
@@ -17,6 +17,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added new image metric `UniversalImageQualityIndex` ([#824](https://github.com/PyTorchLightning/metrics/pull/824))
 
 
+- Added smart update of `MetricCollection` ([#718](https://github.com/PyTorchLightning/metrics/pull/709))
+
```
Comment (Borda) [inline CHANGELOG.md:21]:
```suggestion


```

File: docs/source/pages/overview.rst [hunk 12]
```diff
@@ -292,6 +292,14 @@ inside your LightningModule
     have the same call signature. If this is not the case, input that should be
     given to different metrics can given as keyword arguments to the collection.
 
+An additional advantage of using the ``MetricCollection`` object is that it will
+automatically try to reduce the computations needed by finding groups of metrics
+that share the same underlying metric state. If such metrics are found only one
```
Comment (Borda) [inline docs/source/pages/overview.rst:297]:
```suggestion
that share the same underlying metric state. If such a group of metrics is found only one
```
can we also say that only the first os updated to reduce number one :)

File: docs/source/pages/overview.rst [hunk 13]
```diff
@@ -292,6 +292,14 @@ inside your LightningModule
...
+of them are actually updated and the updated state will be broadcastet to the rest
+of the metrics. In the example above, this will lead to a 2x-3x lower computational
```
Comment (Borda) [inline docs/source/pages/overview.rst:299]:
```suggestion
of the metrics within the group. In the example above, this will lead to a 2x-3x lower computational
```

File: docs/source/pages/overview.rst [hunk 14]
```diff
@@ -292,6 +292,14 @@ inside your LightningModule
...
+cost compared to disabling this feature. See the `compute_groups` argument
```
Comment (Borda) [inline docs/source/pages/overview.rst]:
```suggestion
cost compared to disabling this feature. See the ``compute_groups`` argument
```

File: docs/source/pages/overview.rst (hunk 10 above)
Comment (Borda) [reply docs/source/pages/overview.rst]: > high cost upfront

how much high, this sounds like rather skip it...

Comment (Borda) [reply docs/source/pages/overview.rst]: I see so the overhead is fixed, so it may not pay off for very small scale training of about 100 batches but widely pays of for mid and large scale trainings

Comment (Borda) [reply docs/source/pages/overview.rst]: I think that this is the best place we want to show a graph which we can under as matplotlib figure
see: https://github.com/PyTorchLightning/pytorch-lightning/issues/5630

File: docs/source/pages/overview.rst [hunk 15]
```diff
@@ -292,6 +292,14 @@ inside your LightningModule
...
+for more info.
+
```
Comment (Borda) [inline docs/source/pages/overview.rst:303]: maybe add some examples of such groups, so for classification, it is quite natural, but it may be hard for audio...

File: tests/bases/test_collections.py [hunk 16]
```diff
@@ -278,3 +277,87 @@ def compute(self):
...
+    target = torch.randint(3, (10,))
+    # compute groups should kick in here
+    m.update(preds, target)
+    m2.update(preds, target)
+
+    # compare results for correctness
+    res_cg = m.compute()
+    res_without_cg = m2.compute()
+    for key in res_cg.keys():
+        assert torch.allclose(res_cg[key], res_without_cg[key])
```
Comment (Borda) [inline tests/bases/test_collections.py:339]: shall we include some time measure to be sure that the update does not explode...
so for example one test with 3x the same classic metric and reasonable nb batches to be sure that without > with

File: torchmetrics/collections.py [hunk 17]
```diff
@@ -45,6 +44,14 @@ class name as key for the output dict.
 
         postfix: a string to append after the keys of the output dict
 
+        compute_groups:
+            By default the MetricCollection will try to reduce the computations needed for the metrics in the collection
+            by checking if they belong to the same *compute group*. All metrics in a compute group share the same metric
```
Comment (Borda) [inline torchmetrics/collections.py:49]:
```suggestion
            by checking if they belong to the same **compute group**. All metrics in a compute group share the same metric
```

File: torchmetrics/collections.py [hunk 18]
```diff
@@ -45,6 +44,14 @@ class name as key for the output dict.
...
+            state and are therefore only different in their compute step e.g. accuracy, precision and recall can all be
+            computed from the true positives/negatives and false positives/nagatives. By default this argument is `True`
```
Comment (Borda) [inline torchmetrics/collections.py:51]:
```suggestion
            computed from the true positives/negatives and false positives/negatives. By default, this argument is ``True``
```

**Comment by justusschock** [comment]
Seems sphinx doesn't know `:param:`, should we try `:attr:` instead? Param may only work when being within this function...

**Comment by SkafteNicki** [comment]
@justusschock I changed the text to not include a hyperlink but instead just reference the class docs that are just below the sentence

File: CHANGELOG.md [hunk 19]
```diff
@@ -17,6 +17,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added new image metric `UniversalImageQualityIndex` ([#824](https://github.com/PyTorchLightning/metrics/pull/824))
 
 
+- Added smart update of `MetricCollection` ([#718](https://github.com/PyTorchLightning/metrics/pull/709))
```
Comment (Borda) [inline CHANGELOG.md:20]:
```suggestion
- Added smart update of `MetricCollection` ([#709](https://github.com/PyTorchLightning/metrics/pull/709))
```

File: docs/source/pages/overview.rst [hunk 20]
```diff
@@ -300,6 +300,17 @@ inside your LightningModule
     have the same call signature. If this is not the case, input that should be
     given to different metrics can given as keyword arguments to the collection.
 
+An additional advantage of using the ``MetricCollection`` object is that it will
+automatically try to reduce the computations needed by finding groups of metrics
+that share the same underlying metric state. If such a group of metrics is found only one
+of them is actually updated and the updated state will be broadcasted to the rest
+of the metrics within the group. In the example above, this will lead to a 2x-3x lower computational
+cost compared to disabling this feature. However, this speedup comes with a high cost upfront, where
```
Comment (Borda) [inline docs/source/pages/overview.rst:308]:
```suggestion
cost compared to disabling this feature. However, this speedup comes with a fixed cost upfront, where
```
the high sounds scary...

File: docs/source/pages/overview.rst [hunk 21]
```diff
@@ -300,6 +300,17 @@ inside your LightningModule
...
+the state-groups have to be determined after the first update. This cost can be significant for very
```
Comment (Borda) [inline docs/source/pages/overview.rst:309]:
```suggestion
the state-groups have to be determined after the first update. This overhead can be significantly higher then gains speed-up for very
```

File: docs/source/pages/overview.rst [hunk 22]
```diff
@@ -300,6 +300,17 @@ inside your LightningModule
...
+low steps (approx. up to 100) but still leads to an overall speedup for everything beyond that.
```
Comment (Borda) [inline docs/source/pages/overview.rst:310]:
```suggestion
a low number of steps (approx. up to 100) but still leads to an overall speedup for everything beyond that.
```

File: docs/source/pages/overview.rst [hunk 23]
```diff
@@ -300,6 +300,17 @@ inside your LightningModule
...
+In case the groups are known beforehand, these can also be set manually to avoid the cost of the
```
Comment (Borda) [inline docs/source/pages/overview.rst:311]:
```suggestion
In case the groups are known beforehand, these can also be set manually to avoid this extra cost of the
```

**Review by Borda — APPROVED** [review]

**Comment by tchaton** [comment]
@SkafteNicki Beautiful addition !
